Common/src/main/java/com/blamejared/crafttweaker/api/tag/CraftTweakerTagRegistry.java

```diff
@@ -0,0 +1,299 @@
+import com.blamejared.crafttweaker.api.util.InstantiationUtil;
+import com.blamejared.crafttweaker.platform.Services;
+import com.google.common.base.Suppliers;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.Tag;
+import net.minecraft.tags.TagManager;
+import net.minecraft.tags.TagNetworkSerialization;
+import org.openzen.zencode.java.ZenCodeGlobals;
+import org.openzen.zencode.java.ZenCodeType;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+@ZenRegister
+@ZenCodeType.Name("crafttweaker.api.tag.TagManager")
+public final class CraftTweakerTagRegistry {
+    
+    private static final Supplier<Set<String>> SERVER_ONLY_FOLDERS = Suppliers.memoize(() -> Services.REGISTRY.serverOnlyRegistries()
+            .stream()
+            .map(CraftTweakerTagRegistry.INSTANCE::makeTagFolder)
+            .map(ResourceLocation::tryParse)
+            .filter(Objects::nonNull)
+            .map(ResourceLocation::toString)
+            .collect(Collectors.toSet()));
+    
+    public static final String GLOBAL_NAME = "tags";
+    
+    @ZenCodeGlobals.Global(GLOBAL_NAME)
+    public static final CraftTweakerTagRegistry INSTANCE = new CraftTweakerTagRegistry();
+    
+    private CraftTweakerTagRegistry() {}
+    
+    private final Map<ResourceKey<? extends Registry<?>>, ITagManager<?>> registeredManagers = new HashMap<>();
+    private final Set<ResourceKey<? extends Registry<?>>> knownManagers = new HashSet<>();
+    private final Set<ResourceKey<? extends Registry<?>>> knownManagersView = Collections.unmodifiableSet(knownManagers);
+    
+    /**
+     * Adds a new {@link ITagManager} to the registry.
+     *
+     * @param cls The class of the {@link ITagManager} to add.
+     *
+     * @return The {@link ITagManager} that was added.
+     */
+    public <T> ITagManager<?> addManager(Class<? extends ITagManager<?>> cls) {
+        
+        ITagManager<?> manager = InstantiationUtil.getOrCreateInstance(cls);
+        Objects.requireNonNull(manager, "Error while creating tag manager from class: '" + cls + "'! Make sure it has a default constructor or a public static INSTANCE field!");
+        return addManager(manager);
+    }
+    
+    /**
+     * Adds a new {@link ITagManager} to the registry.
+     *
+     * @param manager The {@link ITagManager} to add.
+     *
+     * @return The {@link ITagManager} that was added.
+     */
+    public <T> ITagManager<?> addManager(ITagManager<?> manager) {
+        
+        registeredManagers.put(manager.resourceKey(), manager);
+        if(manager.getClass().equals(KnownTagManager.class)) {
```

Comment:
Class equality vs. assignableFrom (should manager.getClass() be allowed to point to a subclass of KTM or be exactly KTM)

```suggestion
        if(KnownTagManager.isAssignableFrom(manager.getClass())) {
```

Reply:
In this case we specifically want KTM since it has type parameters, subclasses of KTM aren't allowed type params (like EntityTypeTagManager)